Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,355 @@
+      maximumPingsWithoutData: configuration.maximumPingsWithoutData,
+      minimumSentPingIntervalWithoutData: configuration.minimumSentPingIntervalWithoutData,
+      minimumReceivedPingIntervalWithoutData: configuration.minimumReceivedPingIntervalWithoutData,
+      maximumPingStrikes: configuration.maximumPingStrikes
+    )
+  }
+
+  /// The ping handler.
+  var pingHandler: PingHandler
+
+  /// The scheduled task which will ping.
+  var scheduledPing: RepeatedTask? = nil
+
+  /// The scheduled task which will close the connection.
+  var scheduledClose: Scheduled<Void>? = nil
+
+  func close(context: ChannelHandlerContext) {
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+protocol _ChannelKeepaliveHandler: ChannelInboundHandler where OutboundOut == HTTP2Frame, InboundIn == HTTP2Frame {
+  var pingHandler: PingHandler { get set }
+  var scheduledPing: RepeatedTask? { get set }
+  var scheduledClose: Scheduled<Void>? { get set }
+  func close(context: ChannelHandlerContext)
+}
+
+extension _ChannelKeepaliveHandler {
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    if event is NIOHTTP2StreamCreatedEvent {
+      self.perform(action: self.pingHandler.streamCreated(), context: context)
+    } else if event is StreamClosedEvent {
+      self.perform(action: self.pingHandler.streamClosed(), context: context)
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    switch self.unwrapInboundIn(data).payload {
+    case let .ping(pingData, ack: ack):
+      self.perform(action: self.pingHandler.read(pingData: pingData, ack: ack), context: context)
+    default:
+      break
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  func handlerRemoved(context: ChannelHandlerContext) {
+    self.cancelScheduledPing()
+    self.cancelScheduledTimeout()
+  }
+
+  private func perform(action: PingHandler.Action, context: ChannelHandlerContext) {
+    switch action {
+    case let .schedulePing(delay, timeout):
+      self.schedulePing(delay: delay, timeout: timeout, context: context)
+    case .cancelScheduledTimeout:
+      self.cancelScheduledTimeout()
+    case let .reply(payload):
+      self.send(payload: payload, context: context)
+    case .none:
+      break
+    }
+  }
+
+  private func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext) {
+    let frame = self.wrapOutboundOut(.init(streamID: .rootStream, payload: payload))
+    context.writeAndFlush(frame, promise: nil)
+  }
+
+  private func schedulePing(delay: TimeAmount, timeout: TimeAmount, context: ChannelHandlerContext) {
+    guard delay != .nanoseconds(Int64.max) else { return }
+
+    self.scheduledPing = context.eventLoop.scheduleRepeatedTask(initialDelay: delay, delay: delay) { _ in
+      self.perform(action: self.pingHandler.pingFired(), context: context)
+      self.scheduleClose(timeout: timeout, context: context)
```

Comment:
Hmm, this seems strange. We'll unconditionally schedule a close here, but that does not include cancelling any previously scheduled close. Is it our expectation that no previously scheduled close exists? What's the intended behaviour?

I find this line of code surprising enough that I think we should at least comment what the intended behaviour is.

Reply:
The scheduled close is getting cancelled as soon as we get a pong back. I think it makes sense that `timeout` value is smaller than `interval` so we shouldn't have a situation where a new ping is generated before either a pong is received or the timeout is executed. I see that this is not documented and I realize now that we might have situations where a `timeout` might be greater than `interval`...

Reply:
Yeah, if we want that to be a criterion that's fine, but we should a) add code to enforce that invariant, and b) comment here that that invariant is why we don't need to worry about this, and c) add an assertion for debug mode that ensures that invariant is upheld.

Reply:
Would you rather have a failable initializer for both `ClientConnectionKeepalive` and `ServerConnectionKeepalive` which would guarantee a valid configuration?

Reply:
Nevermind, carrying optional isn't fun.

Reply:
Looking at the go implementation, `timeout` can be greater than `interval`, so we should respect that too. 
https://godoc.org/google.golang.org/grpc/keepalive

Reply:
> Looking at the go implementation, `timeout` can be greater than `interval`, so we should respect that too.
> https://godoc.org/google.golang.org/grpc/keepalive

Unless I missed something, it's not clear to me whether that is or is not the case from the linked docs.

Are there cases where it would make sense for the `timeout` to be larger than the `interval`?

Reply:
I agree that a timeout larger than interval doesn't make sense. I'm fine making sure this doesn't happen. What's the best way to validate the configuration?

Reply:
The answer there is usually to `precondition` on a valid configuration and express what the validity criteria are in the docs. As these values are unlikely to be chosen based on user-supplied data, a `precondition` is rarely going to be an issue here.

Reply:
> Yeah, if we want that to be a criterion that's fine, but we should a) add code to enforce that invariant, and b) comment here that that invariant is why we don't need to worry about this, and c) add an assertion for debug mode that ensures that invariant is upheld.

b) and c) still need addressing.